Update 3D lights and shadows documentation for 4.0 (#6355)
godotengine/godot-docs: tutorials/3d/lights_and_shadows.rst

- **Cull Mask:** Objects that are in the selected layers below will be affected by this light.
  Note that objects disabled via this cull mask will still cast shadows.
  If you don't want disabled objects to cast shadows, adjust the **Cast Shadow**
  property on the GeometryInstance3D to the desired value.

.. seealso::

    See :ref:`doc_physical_light_and_camera_units` if you wish to use real world
    units to configure your lights' intensity and color temperature.

Light number limits
-------------------

When using the Forward+ renderer, Godot uses a *clustering* approach for
real-time lighting. As many lights as desired can be added (as long as
performance allows). However, there's still a default limit of 512 *clustered
elements* that can be present in the current camera view. A clustered element is
an omni light, a spot light, a :ref:`decal <doc_decals>` or a :ref:`reflection
probe <doc_reflection_probes>`. This limit can be increased by adjusting the
**Rendering > Limits > Cluster Builder > Max Clustered Elements** advanced
project setting.

When using the Forward Mobile renderer, there is a limitation of 8 OmniLights +

> 8 SpotLights per mesh resource. This currently cannot be changed.

Reviewer: There is also a limit of 256 lights per scene, also not configurable

Author: Is this 256 OmniLights + 256 SpotLights, or 256 OmniLights/SpotLights (meaning you can't have more than 128 of each if you want a 50-50 split)?

Reviewer: You can have 256 Omnilights and 256 Spotlights (so up to a total of 512)

When using the Compatibility renderer, up to 8 OmniLights + 8 SpotLights can be
rendered per mesh resource. This limit can be increased in the advanced Project
Settings by adjusting **Rendering > Limits > OpenGL > Max Renderable Lights**
and/or **Rendering > Limits > OpenGL > Max Lights Per Object** at the cost of
performance and longer shader compilation times. The limit can also be decreased
to reduce shader compilation times and improve performance slightly.

> Additionally, only one DirectionalLight can be visible at a time.

Reviewer: Multiple DirectionalLight3Ds can be visible at the same time. You can just remove this whole line

Author: This is a leftover I forgot to remove – I can confirm multiple lights can be seen when using OpenGL.

not reach occluded areas), but it can incur a bigger performance cost.
There is a list of generic shadow parameters, each also has a specific function:

- **Enabled:** Check to enable shadow mapping in this light.
- **Opacity:** Areas occluded are darkened by this opacity factor. Shadows are
  fully opaque by default, but this can be changed to make shadows translucent
  for a given light.
- **Bias:** When this parameter is too small, self shadowing occurs. When too
  large, shadows separate from the casters. Tweak to what works best for you.
- **Contact:** Performs a short screen-space raycast to reduce the gap generated
  by the bias. Contact shadows are only available when using the GLES3 backend.
- **Reverse Cull Face:** Some scenes work better when shadow mapping is rendered
  with face-culling inverted.

> 

Reviewer 2: Several things about the above section.
- Contact is no longer a parameter at least in beta 4
- Transmittance bias, normal bias and blur are all missing from this list.

will make the light fade out slower and shadows appear blurrier. This can be
used to simulate area lights to an extent. This kind of shadow is expensive,
so check the recommendations in :ref:`doc_lights_and_shadows_pcss_recommendations`
if setting this value above ``0.0`` on lights with shadows enabled.

Omni shadow mapping
^^^^^^^^^^^^^^^^^^^

Omni light shadow mapping is relatively straightforward. The main issue that
needs to be considered is the algorithm used to render it.

Omni Shadows can be rendered as either **Dual Paraboloid** or **Cube** mapped.
**Dual Parabolid** renders quickly, but can cause deformations, while **Cube**
is more correct, but slower. The default is **Cube**, but consider changing it
to **Dual Parabolid** for lights where it doesn't make much of a visual
difference.

.. image:: img/lights_and_shadows_dual_parabolid_vs_cubemap.webp

If the objects being rendered are mostly irregular and subdivided, Dual
Paraboloid is usually enough. In any case, as these shadows are cached in a
shadow atlas (more on that at the end), it may not make a difference in
performance for most scenes.

Omni lights with shadows enabled can make use of projectors. The projector

> texture will *multiply* the lights's color by the color at a given point on the

Reviewer 2: ```suggestion
texture will *multiply* the light's color by the color at a given point on the
```

car lights, reflectors, spots, etc. This type of light is also attenuated towards the
opposite direction it points to.

Spot lights share the same **Range**, **Attenuation** and **Size** as OmniLight3D,
and add two extra parameters:

- **Angle:** The aperture angle of the light.
- **Angle Attenuation:** The cone attenuation, which helps soften the cone borders.

Spot shadow mapping
^^^^^^^^^^^^^^^^^^^

Spots feature the same parameters as omni lights for shadow mapping. Rendering
spot shadow maps is significantly faster compared to omni lights, as only one
shadow texture needs to be rendered (instead of rendering 6 faces, or 2 in dual
parabolid mode).

Spot lights with shadows enabled can make use of projectors. The projector

> texture will *multiply* the lights's color by the color at a given point on the

Reviewer 2: ```suggestion
texture will *multiply* the light's color by the color at a given point on the
```

fewer shadows. This will allow each shadow to be rendered at a higher resolution.

.. _doc_lights_and_shadows_shadow_filter_mode:

Shadow filter mode
^^^^^^^^^^^^^^^^^^

Several shadow map quality settings can be chosen here. The default **Soft Low**
is a good balance between performance and quality for scenes with detailed
textures, as the texture detail will help make the dithering pattern less noticeable.

However, in projects with less detailed textures, the shadow dithering pattern
may be more visible. To hide this pattern, you can either enable
:ref:`doc_3d_antialiasing_taa`, :ref:`doc_3d_antialiasing_fxaa`, or increase the
shadow filter quality to **Soft Medium** or higher.

The **Soft Very Low** setting will automatically decrease shadow blur to make
artifacts from the low sample count less visible. Conversely, the **Soft High**
and **Soft Ultra** settings will automatically increase shadow blur to better
make use of the increased sample count.

16-bits versus 24-bit
^^^^^^^^^^^^^^^^^^^^^

By default, Godot uses 16-bit depth textures for shadow map rendering. This is
recommended in most cases as it performs better without a noticeable difference
in quality.

> If **16 Bits** is disabled, 24-bit shadow rendering will be used instead. This

Reviewer: ```suggestion
If **16 Bits** is disabled, 32-bit depth textures will be used instead. This
```

Author: I thought we were using 24-bit depth instead, but I guess it depends on hardware support (we should be using 32-bit everywhere nowadays, anyway).
